src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -881,42 +883,206 @@ protected virtual void Dispose(bool disposing)
+        /// <para>Important performance considerations:</para>
+        /// <para>The <c>SetCustomProperty(..)</c> and the corresponding <c>GetCustomProperty(..)</c> methods
+        /// are performance-optimized for scenarios where the number of different property names is small,
+        /// and the number of property Get operations strongly dominates the number of Set operations.</para>
+        /// <para>When the number of  property name-value pairs is small, the data is held in an array.
+        /// Look-up is performed via a linear scan of the array, however, using very fast comparisons.
+        /// Updates use optimistic concurrency and perform a full array copy.
+        /// Once the number of  property name-value pairs crosses a certain threshold,
+        /// a concurrent dictionary is used to hold the data instead of the array.
+        /// Look-up are performed in an amortized constant time, however, the required book-keeping may
+        /// have a significant performance impact if properties are accessed very frequently.</para>
+        /// <para>If you need a large number of custom properties, consider using a custom ActivityProperties
+        /// class, instead of using many properties entries.</para>
+        /// <para>For example, the following code is less performant than the code below.</para>
+        /// <para>Less performant code:</para>
+        /// <code>
+        ///   Activity a = new Activity("a");
+        ///   a.SetCustomProperty("Your.Company.SpecialId", someGuid);
+        ///   a.SetCustomProperty("Your.Company.RelatedActivity", someActivity);
+        ///   a.SetCustomProperty("Your.Company.ExtraName", "Lorem Ipsum");
+        ///   . . .
+        ///   string extraName = (string) a.GetCustomProperty("Your.Company.ExtraName");
+        /// </code>
+        /// <para>Equivalent faster code (number of property set/get operations is reduced):</para>
+        /// <code>
+        ///   internal class ProprietaryTracingInfo
+        ///   {
+        ///       public Guid SpecialId { get; set; }
+        ///       public Activity RelatedActivity { get; set; }
+        ///       public string ExtraName { get; set; }
+        ///   }
+        ///   . . .
+        ///   Activity a = new Activity("a");
+        ///   a.SetCustomProperty("Your.Company.ProprietaryTracingInfo", new ProprietaryTracingInfo() {  ExtraName = "Lorem Ipsum" });
+        ///   . . .
+        ///   string extraName = ((ProprietaryTracingInfo) a.GetCustomProperty("Your.Company.ExtraName")).ExtraName;
+        /// </code>
+        /// </remarks>
         /// <param name="propertyName"> The name to associate the value with.<see cref="OperationName"/></param>
         /// <param name="propertyValue">The object to attach and map to the property name.</param>
         public void SetCustomProperty(string propertyName, object? propertyValue)
         {
-            if (_customProperties == null)
+            object? currProps = _customProperties;
```

Comment:
This is getting pretty sizable, we probably need an end-to-end benchmark to show this improvement is meaningful. One potential issue is that a microbenchmark may show clear improvement but an end-to-end won't. Increased code size (in particular increased demand for branch predictor cache cells) may mean that even if this snippet speeds up other parts of the code slow down.

I also think it would be worthwhile doing a simple implementation using a (non-concurrent) Dictionary and a monitor lock. I suspect it may be a good bit faster than the ConcurrentDictionary and makes a more reasonable baseline for the perf obtainable with a small code/low complexity solution.

Reply:
I agree this is really getting complicated. I would start first with normal Dictionary with a lock (not using concurrent dictionary nor array) and look at the benchmark result. And then we can improve more from there. I think the implementation can  be simplified. I understand the goal of enhancing the performance but in same time the code maintenance is important too. 
I understand this is a draft so you didn't even compiled the code. 


---
In reply to: [458003983](https://github.com/dotnet/runtime/pull/39690#discussion_r458003983) [](ancestors = 458003983)

Reply:
Thanks for the feedback.

@noahfalk , any particular E2E benchmark that you had in mind? A suggestion would be to measure an OTel Tracer prototype with different Activity implementations discussed here.

I used a Concurrent Dictionary over a normal + lock because it only matters once the fast access array is exceeded.
I recall from the 4.5 age that a Concurrent Dictionary used almost for Reading was only marginally slower than a normal Dictionary. Is it not the case?

Reply:
> I recall from the 4.5 age that a Concurrent Dictionary used almost for Reading was only marginally slower than a normal Dictionary. Is it not the case?

Last I recall Dictionary was more than twice as fast but the BCL team and JIT are always optimizing things over time so its hard to say if my info is up-to-date. I can give guesses about how things perform, but the way to know is to test it.

Reply:
> @noahfalk , any particular E2E benchmark that you had in mind? A suggestion would be to measure an OTel Tracer prototype with different Activity implementations discussed here.

I think the closer we can get to real world the better so that would be a great benchmark IMO.